src/ucs/stats/stats.h

```diff
@@ -14,15 +15,87 @@
 
 #include <ucs/sys/compiler_def.h>
 #include <ucs/sys/stubs.h>
+#include <ucs/stats/stats_fwd.h>
 
 BEGIN_C_DECLS
 
 /** @file stats.h */
 
+/* The maximum size of a class name */
+#define UCS_STATS_CLASS_NAME_MAX 50
+
+/* The allocation name of the database */
+#define UCS_STATS_CLSID_TO_NAME_DB_NAME "stats_clsid2name_db"
+
+/* The allocation name of the aggregate-sum counters names database */
+#define UCS_STATS_AGGRGT_SUM_CNT_NAMES_DB "stats_cnt_name_db"
+
+/* Realloc size (i.e. how many elements to add during realloc()) */
+#define UCS_STATS_AGG_SUM_COUNTERS_REALLOC_SZ 32
+
+
+typedef struct {
+    /* Class name */
+    char     class_name[UCS_STATS_CLASS_NAME_MAX];
+
+    /* Class base offset in the counters aggregate-sum list */
+    unsigned aggregate_sum_class_offset;
+
+} ucs_stats_aggrgt_sum_clsid_to_name_db_t;
+
+
+typedef struct {
+    /* Counter class name */
+    const char *class_name;
+
+    /* Counter name */
+    const char *name;
+
+} ucs_stats_aggrgt_sum_counter_name_t;
+
+
 void ucs_stats_init();
 void ucs_stats_cleanup();
 void ucs_stats_dump();
 int ucs_stats_is_active();
+
+/**
+ * A UCX statistics API function to return the aggregate-sum of all counters
+ * from all workers/interfaces for a compact statistics trace.
+ * Used for Score-P UCX plugin to collect a compact trace of the counters.
+ *
+ * @param [out] aggregate_sum_counters      The aggregate-sum of all counters of each class/type.
+ * @param [in]  aggregate_sum_counters_size The number of elements in aggregate_sum_counters.
+ * @param [out] n_counters                  The actual number of counters in the output buffer.
+ * @param [out] clsid_to_name_db            The conversion database from clsid to name.
+ *
+ * Complexity: O(n) / recursive.
+ *
+ * Returns: None.
+ *
+ * Remarks: The function calls a recursive function.
+ */
+void ucs_stats_aggregate_sum(
+        ucs_stats_counter_t *aggregate_sum_counters,
+        size_t aggregate_sum_counters_size, size_t *n_counters,
+        const ucs_stats_aggrgt_sum_clsid_to_name_db_t **clsid_to_name_db);
+
+/**
+ * A UCX statistics API to get the names of counters on the last call to
+ * ucs_stats_aggregate_sum()
+ *
+ * @param [out] names_db The aggregate-sum counters names database
+ * @param [out] size     number of counters in database
+ *
+ * Complexity: O(1)
+ *
+ * Returns: None.
+ *
+ */
+void ucs_stats_aggregate_sum_names_db_get(
```

Comment:
1. what if ucs_stats_aggregate_sum_names_db_get is called before ucs_stats_aggregate_sum, seems it will return empty array?
2. what guarantees the global counter db did not change between call to ucs_stats_aggregate_sum() and ucs_stats_aggregate_sum_names_db_get()?

Reply:
1. Yes, I don't see a problem with that.
2. Even if the global counter database changes between the calls, there's no problem since the snapshot only has class_id's from the first call and the change will not cause a bug or crash.